fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,527 @@
+    @Deprecated
+    public byte[] getContinuation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public RecordCursor.NoNextReason getNoNextReason() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() {
+        if (hasNextFuture != null) {
+            hasNextFuture.cancel(false);
+        }
+    }
+
+    //form a continuation that allows us to restart statistics aggregation from where we left off
+    private class SizeStatisticsCollectorCursorContinuation implements RecordCursorContinuation {
+        @Nonnull
+        private final RecordCursorResult<KeyValue> currentKvResult;
+        @Nonnull
+        private final Function<byte[], RecordCursorProto.SizeStatisticsContinuation> continuationFunction;
+        @Nonnull
+        private final boolean isEnd = false;
+        @Nullable
+        private byte[] cachedBytes;
+
+        private SizeStatisticsCollectorCursorContinuation(RecordCursorResult<KeyValue> currentKvResult) {
+            this.cachedBytes = null;
+            this.currentKvResult = currentKvResult;
+
+            //defer forming bytes until requested
+            continuationFunction = b -> {
+                RecordCursorProto.SizeStatisticsPartialResults.Builder pr;
+                if (finalResultsEmitted == false) {
+                    //more to aggregate put partial results in continuation
+                    pr = RecordCursorProto.SizeStatisticsPartialResults.newBuilder().setKeyCount(sizeStatisticsResults.keyCount).setKeySize(sizeStatisticsResults.keySize).setMaxKeySize(sizeStatisticsResults.maxKeySize).setValueSize(sizeStatisticsResults.valueSize).setMaxValueSize(sizeStatisticsResults.maxValueSize);
+                    List<Long> bucketList = Arrays.stream(sizeStatisticsResults.getSizeBuckets()).boxed().collect(Collectors.toList());
+                    pr.addAllSizeBuckets(bucketList);
+                    return RecordCursorProto.SizeStatisticsContinuation.newBuilder().setPartialResults(pr.build()).setContinuation(ByteString.copyFrom(b)).build();
+                } else {
+                    //nothing more to aggregate
+                    return RecordCursorProto.SizeStatisticsContinuation.newBuilder().setContinuation(ByteString.copyFrom(RecordCursorEndContinuation.END.toBytes())).build();
+                }
+            };
+        }
+
+        @Nullable
+        @Override
+        public byte[] toBytes() {
+            if (isEnd()) {
+                return null;
+            } else {
+                //form bytes exactly once
+                if (this.cachedBytes == null) {
+                    this.cachedBytes = continuationFunction.apply(this.currentKvResult.getContinuation().toBytes()).toByteArray();
+                }
+                return cachedBytes;
+            }
+        }
+
+        @Override
+        @Nonnull
+        public boolean isEnd() {
+            return isEnd;
```

Comment:
Are there other continuations besides `RecordCursorContinuation.END` for which `isEnd` returns `true`? Maybe I'm misremembering, but I thought that was the only one.

Maybe this needs a static function that returns `RecordCursorContinuation` and returns `RecordCursorContinuation.END` if it's at the end and one of these continuations otherwise?

Reply:
Yeah, I am confused about this in general. Seems any cursor should return RecordCursorContinuation.END  when input is exhausted, and a continuation where isEnd() is false otherwise. The knowledge of which one to use is probably somewhere in the onNext() method of a given cursor. Seems this decision logic would rely on instance variables so not sure how a static method would make the decision. Anyway, if this particular method is used here for this cursor ifEnd() will be false.

Reply:
Maybe a static initializer isn't necessary, but the point was more that instead of calling the constructor directly, there's a like `createContinuation` method that returns `RecordCursorContinuation.END` if we're at the end and one of these continuations otherwise.

Reply:
The code is immediately returning in onNext() with RecordCursorResult.exhausted()) when it is called after results have already been emitted. I think you are suggesting to instead change that to something like .. 
RecordCursorResult.withoutNextValue(getSizeStatisticsCollectorCursorContinuation(resultKv, sizeStatisticsResults, <some other state>), NoNextReason.SOURCE_EXHAUSTED); 
The issue is the <some other state> would need to indicate that this call is for an end continuation based on the context (i.e. where it is in the onNext() method.) Can make such a change but would like to understand the benefit better.

Reply:
The benefit would be that you could return the statically allocated `RecordCursorContinuation.END` object instead of creating a new one when `isEnd` is `true`. A few continuations, like the `ByteArrayContinuation`, work like that, though there are also a few cursors for which `isEnd` might return `true` or `false`.

Looking through this cursor, it appears that `isEnd` is initialized to `false` and never mutated for this continuation object, so I think it actually already does this. I think that's actually the right behavior, but `isEnd` probably shouldn't be a member variable.